contracts/BSCValidatorSet.sol

```diff
@@ -312,7 +319,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     return CODE_OK;
   }
 
-  function getValidators()external view returns(address[] memory) {
+  function shuffle(address[] memory validators, uint256 epochNumber, uint startIdx, uint offset, uint limit, uint modNumber) internal pure {
+    for (uint i = 0; i<limit; i++) {
+      uint random = uint(keccak256(abi.encodePacked(epochNumber, startIdx+i))) % modNumber;
+      if ( (startIdx+i) != (offset+random) ) {
+        address tmp = validators[startIdx+i];
+        validators[startIdx+i] = validators[offset+random];
+        validators[offset+random] = tmp;
+      }
+    }
+  }
+
+  function getMiningValidators() public view returns(address[] memory) {
+    uint256 _maxNumOfWorkingCandidates = maxNumOfWorkingCandidates;
+    uint256 _numOfCabinets = numOfCabinets;
+    if (_numOfCabinets == 0 ){
+      _numOfCabinets = INIT_NUM_OF_CABINETS;
+    }
+
+    address[] memory validators = getValidators();
```

Comment:
Shall we ensure the validatorSet is sorted in checkValidatorSet()? If the validatorSet is passed not sorted, we should reject it in the first place.

Reply:
It is guaranteed by `BNB beacon chain`.